src/xref2/ref_tools.ml

```diff
@@ -61,183 +85,409 @@ let module_type_lookup_to_signature_lookup :
+
+  let in_env env name : t option =
+    Env.lookup_datatype_by_name (UnitName.to_string name) env >>= function
+    | `Type _ as e -> Some (of_element env e)
+    | _ -> None
+
+  let in_signature _env ((parent', parent_cp, sg) : signature_lookup_result)
+      name : t option =
+    let sg = Tools.prefix_signature (parent_cp, sg) in
+    Find.datatype_in_sig sg (TypeName.to_string name) >>= fun t ->
+    Some (`Type (parent', name), t)
+end
+
+module T = struct
+  (** Type *)
+
+  type t = type_lookup_result
+
+  let in_env env name : type_lookup_result option =
+    Env.lookup_datatype_by_name (UnitName.to_string name) env >>= function
+    | `Type (id, t) -> Some (`T (`Identifier id, t))
+    | `Class _ as e -> Some (`C (CL.of_element env e))
+    | `ClassType _ as e -> Some (`CT (CT.of_element env e))
+
+  (* Don't handle name collisions between class, class types and type decls *)
+  let in_signature _env ((parent', parent_cp, sg) : signature_lookup_result)
+      name : type_lookup_result option =
+    let sg = Tools.prefix_signature (parent_cp, sg) in
+    Find.type_in_sig sg name >>= function
+    | `T t -> Some (`T (`Type (parent', TypeName.of_string name), t))
+    | `C c -> Some (`C (`Class (parent', ClassName.of_string name), c))
+    | `CT ct ->
+        Some (`CT (`ClassType (parent', ClassTypeName.of_string name), ct))
+end
+
+module V = struct
+  (** Value *)
+
+  type t = value_lookup_result
+
+  let in_env env name : t option =
+    Env.lookup_value_by_name (UnitName.to_string name) env >>= function
+    | `Value (id, _x) -> return (`Identifier id)
+    | `External (id, _x) -> return (`Identifier id)
+
+  let of_component _env ~parent_ref name : t option =
+    Some (`Value (parent_ref, ValueName.of_string name))
+
+  let external_of_component _env ~parent_ref name : t option =
+    (* Should add an [`External] reference ? *)
+    Some (`Value (parent_ref, ValueName.of_string name))
+
+  let in_signature _env ((parent', _, sg) : signature_lookup_result) name :
+      t option =
+    Find.opt_value_in_sig sg (ValueName.to_string name) >>= fun _ ->
+    Some (`Value (parent', name))
+end
+
+module L = struct
+  (** Label *)
 
-and resolve_label_parent_reference :
+  type t = Resolved.Label.t
+
+  let in_env env name : t option =
+    Env.lookup_label_by_name (UnitName.to_string name) env
+    >>= fun (`Label id) -> Some (`Identifier id)
+
+  let in_page _env (`Page (_, p)) name : t option =
+    try Some (`Identifier (List.assoc name p)) with Not_found -> None
+
+  let of_component _env ~parent_ref name : t option =
+    Some
+      (`Label
+        ((parent_ref :> Resolved.LabelParent.t), LabelName.of_string name))
+
+  let in_label_parent env (parent : label_parent_lookup_result) name : t option
+      =
+    match parent with
+    | `S (p, _, sg) ->
+        Find.opt_label_in_sig sg (LabelName.to_string name) >>= fun _ ->
+        Some (`Label ((p :> Resolved.LabelParent.t), name))
+    | `T _ | `C _ | `CT _ -> None
```

Comment:
This is not implemented on types, classes and type classes yet

Reply:
Anything blocking this?

Reply:
No. I'll work on it.